tools/tplgtool.py

```diff
@@ -599,25 +599,128 @@ def _merge_pcm_list(self, pcm_block_list):
                 merged_pcm_list.append(pcm)
         return merged_pcm_list
 
+    # represent graph with nodes and edges
+    # return values:
+    #   node_widgets -> widget list
+    #   graph_node -> widget name list, used for construct graph
+    #   forward_graph_edges -> adjacency list shows the link between widgets
+    #   backward_graph_edges -> adjacency list shows the reverse link between widgets
+    # The backward_graph_deges is used to backward search the graph, a backward search
+    # can also preceed with forward_graph_edges, but will need more code.
+    def _link_graph(self):
+        graph_nodes = []
+        forward_graph_edges = []
+        backward_graph_edges = []
+        node_widgets = []
+        # traverse original graph list to get a graph node list
+        for graph in self._tplg["graph_list"]:
+            for items in graph:
+                for elem in items:
+                    if elem != '' and elem not in graph_nodes:
+                        graph_nodes.append(elem)
+
+        # traverse original graph list to get a graph edge list
+        # original represention of an edge is a list with three
+        # elements: source control sink.
+        for node in graph_nodes:
+            forward_edges = []
+            backward_edges = []
+            node_widgets.append(self._find_widget_by_name(node))
+            for graph in self._tplg["graph_list"]:
+                for elem in graph:
+                    if node == elem[0] and elem[1] == '':
+                        forward_edges.append(graph_nodes.index(elem[2]))
+                    if node == elem[0] and elem[1] != '':
+                        forward_edges.append(graph_nodes.index(elem[1]))
+                    if node == elem[1]:
+                        forward_edges.append(graph_nodes.index(elem[2]))
+                        backward_edges.append(graph_nodes.index(elem[2]))
+                    if node == elem[2] and elem[1] == '':
+                        backward_edges.append(graph_nodes.index(elem[0]))
+                    if node == elem[2] and elem[1] != '':
+                        backward_edges.append(graph_nodes.index(elem[1]))
+            forward_graph_edges.append(forward_edges)
```

Comment:
I got confused here, what is the `adjacency list shows the link between widgets`? I think we can only have some index here? How to use the index?